fix(android): Fix proxy release for SearchView attached to a TableView.

ypbnv opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-26665

**Description:**
SearchView proxy is now released when the TableView proxy is released instead of when the views are
removed from the screen.

**Test case:**
_app.js_
```js
var win1 = Ti.UI.createWindow({ layout: 'vertical'}),
	tableData = [ {title: 'Apples'}, {title: 'Bananas'}, {title: 'Carrots'}, {title: 'Potatoes'} ],
	button1 = Ti.UI.createButton({ title: 'Remove'}),
	button2 = Ti.UI.createButton({ title: 'Add'}),
	searchView = Ti.UI.Android.createSearchView(),
	table = Ti.UI.createTableView({
		height: '80%',
		search: searchView,
		data: tableData
	});

button1.addEventListener('click', function () {
	win1.remove(table);
});

button2.addEventListener('click', function () {
	win1.add(table);
});

win1.addEventListener('open', function () {
	win1.add(button1);
	win1.add(button2);
	win1.add(table);
});

win1.open();

```
Before this PR removing and adding the table view again will result in a crash.

Note: Unit test is on the way.

ypbnv committed:
fix(android): Fix proxy release for SearchView attached to a TableView.

SearchView proxy is now release when the TableView proxy is released instead of when the views are
removed from the screen.

ypbnv committed:
test(android): Add unit test for persisting SearchView for TableView.

build commented on tests/Resources/ti.ui.tableview.addontest.js:
```diff
@@ -0,0 +1,35 @@
+/* global Ti */
+/* eslint no-unused-expressions: "off" */
+'use strict';
+var should = require('./utilities/assertions');
```
<!--
  0 failure: 
  1 warning:  tests/Resources/t...
  
  
  DangerID: danger-id-default;
  File: tests/Resources/ti.ui.tableview.addontest.js;
  Line: 11;
-->

- :warning: tests/Resources/ti.ui.tableview.addontest.js line 11 – 'should' is assigned a value but never used. (no-unused-vars)

sgtcoolguy replied:
```suggestion
var should = require('./utilities/assertions'); // eslint-disable-line no-unused-vars
```

build commented:
<!--
  1 failure:  Tests have failed...
  0 warning: 
  2 messages
  1 markdown notices
  DangerID: danger-id-default;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Tests have failed, see below for more information.</td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-10568/2/artifact/dist/mobilesdk-8.0.0.v20190104072741-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:x: 2 tests have failed
There are 2 tests failing and 450 skipped out of 2970 total tests.</td>
    </tr>
  </tbody>
</table>

### Tests:

| Classname | Name | Time | Error |
| --- | --- | --- | --- |
| android.os | #freemem() returns a positive Number | 0.001 | expected 0 to be above 0 |
| ios.Titanium.UI.TableView | SearchView persistence | 0 | undefined is not an object (evaluating 'Ti.UI.Android.createSearchView') |

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against b2870165639dafaffc76f8365ca4575c46dead03
</p>

ypbnv committed:
test(android): Fix unit test to target only Android.

sgtcoolguy committed:
Merge branch 'master' into TIMOB-26665

sgtcoolguy commented on android/modules/ui/src/java/ti/modules/titanium/ui/TableViewProxy.java:
```diff
@@ -137,6 +139,13 @@ public void releaseViews()
 	public void release()
 	{
+		// Release search bar proxy if there is one
+		if (getTableView().getNativeView() instanceof RelativeLayout) {
```
Can getTableView() return null?
Also, assign a temp var to hold result of `getTableView().getNativeView()` since you're calling it twice consecutively.

ypbnv replied:
I think not. It either returns the already available table view or creates one. But maybe checking for having PROPERTY_SEARCH assigned a value would be a better solution. We don't really care if the search view was drawn on the screen or not before releasing the proxy it holds. I am on it.

sgtcoolguy commented on android/modules/ui/src/java/ti/modules/titanium/ui/TableViewProxy.java:
```diff
@@ -137,6 +139,13 @@ public void releaseViews()
+		// Release search bar proxy if there is one
+		if (getTableView().getNativeView() instanceof RelativeLayout) {
+			((RelativeLayout) getTableView().getNativeView()).removeAllViews();
+			TiViewProxy searchView = (TiViewProxy) (getProperty(TiC.PROPERTY_SEARCH));
```
no need to wrap getProperty in parens.
```suggestion
			TiViewProxy searchView = (TiViewProxy) getProperty(TiC.PROPERTY_SEARCH);
```

sgtcoolguy commented on android/modules/ui/src/java/ti/modules/titanium/ui/TableViewProxy.java:
```diff
@@ -137,6 +139,13 @@ public void releaseViews()
 	public void release()
 	{
+		// Release search bar proxy if there is one
+		if (getTableView().getNativeView() instanceof RelativeLayout) {
```
Isn't all of this redundant with TiUITableView.java.release()? Why do we need to release the search view in both? Doesn't one "control" the flow and call release on the other one?

sgtcoolguy commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUITableView.java:
```diff
@@ -275,11 +275,11 @@ public void onDestroy(Activity activity)
 			((RelativeLayout) nativeView).removeAllViews();
 			TiViewProxy searchView = (TiViewProxy) (proxy.getProperty(TiC.PROPERTY_SEARCH));
-			searchView.release();
+			searchView.releaseViews();
```
can searchView be null? Or is the layout instanced check an implicit way of "knowing" there's a search view?

ypbnv replied:
Yes, the layout check should only pass if we have a searchView and it is not part of the TableView (Its `searchAsChild` property is set to `false`).

sgtcoolguy commented on tests/Resources/ti.ui.tableview.addontest.js:
```diff
@@ -0,0 +1,35 @@
+			table.removeEventListener('postlayout', removeAndAddTable);
+			win.remove(table);
+			win.add(table);
+			finish();
```
When a function is used as a callback/event listener it will execute "async" and can behave a little strange under unit tests if there's an Error thrown. Typically you'd want to wrap the body of the function where you're doing the work in a `try` block and place the `finish()` call last inside the block - with a `catch(err)` block that calls `finish(err);` to handle error cases.

i.e.
```javascript
function removeAndAddTable() {
  try {
 	table.removeEventListener('postlayout', removeAndAddTable);
 	win.remove(table);
 	win.add(table);
 	finish();
  } catch (err) {
    finish(err);
  }
}
```

Without that I've seen different platforms behave differently - Windows sometimes crashes, the other sometimes just time out the test, leading you to think that there was a timeout/hang (when there wasn't).

ypbnv committed:
refactor(android): Remove redundant clearing of views in TableViewProxy.

Do not remove views in the call of release for TableViewProxy. Also put the unit test in a try-catch
block to avoid breaking the testing sequence.

ypbnv commented:
@sgtcoolguy Updated.

build commented on tests/Resources/ti.ui.tableview.addontest.js:
```diff
@@ -0,0 +1,39 @@
+				win.add(table);
+				finish();
+			} catch (err) {
+				finish (err);
```
<!--
  1 failure:  tests/Resources/t...
  0 warning: 
  
  
  DangerID: danger-id-default;
  File: tests/Resources/ti.ui.tableview.addontest.js;
  Line: 31;
-->
- :no_entry_sign: tests/Resources/ti.ui.tableview.addontest.js line 31 – Unexpected space between function name and paren. (no-spaced-func)

Commit:
test(android): Fix linting for added test.

ypbnv committed:
Merge branch 'master' into TIMOB-26665
